pkg/agent/openflow/pipeline.go

```diff
@@ -454,388 +504,472 @@ func (c *client) DeleteOFEntries(ofEntries []binding.OFEntry) error {
+				LoadToCtMark(ServiceCTMark).
+				LoadToCtMark(HairpinCTMark).
+				CTDone().
+				Done(),
+			// This generates the flow to match the first packet of NodePort / LoadBalancer connection (non-hairpin) initiated
+			// through the Antrea gateway with ConnSNATCTMark, then perform SNAT in SNAT CT zone with the Antrea gateway IP.
+			SNATConntrackCommitTable.ofTable.BuildFlow(priorityLow).
+				Cookie(cookieID).
+				MatchProtocol(ipProtocol).
+				MatchCTStateNew(true).
+				MatchCTStateTrk(true).
+				MatchRegMark(FromGatewayRegMark).
+				MatchCTMark(ConnSNATCTMark).
+				Action().CT(true, SNATConntrackCommitTable.ofTable.GetNext(), f.snatCtZones[ipProtocol]).
+				SNAT(&binding.IPRange{StartIP: f.gatewayIPs[ipProtocol], EndIP: f.gatewayIPs[ipProtocol]}, nil).
+				LoadToCtMark(ServiceCTMark).
+				CTDone().
+				Done(),
+			// This generates the flow to match the subsequent request packets of connection whose first request packet has
+			// been committed in SNAT CT zone, then commit the packets in SNAT CT zone again to perform SNAT.
+			// For example:
+			/*
+				* 192.168.77.1 is the IP address of client.
+				* 192.168.77.100 is the IP address of K8s Node.
+				* 30001 is the NodePort port.
+				* 10.10.0.1 is the IP address of Antrea gateway.
+				* 10.10.0.3 is the IP of NodePort Service Endpoint.
+
+				* packet 1 (request)
+					* client                     192.168.77.1:12345->192.168.77.100:30001
+					* CT zone SNAT 65521         192.168.77.1:12345->192.168.77.100:30001
+					* CT zone DNAT 65520         192.168.77.1:12345->192.168.77.100:30001
+					* CT commit DNAT zone 65520  192.168.77.1:12345->192.168.77.100:30001  =>  192.168.77.1:12345->10.10.0.3:80
+					* CT commit SNAT zone 65521  192.168.77.1:12345->10.10.0.3:80          =>  10.10.0.1:12345->10.10.0.3:80
+					* output
+				  * packet 2 (reply)
+					* Pod                         10.10.0.3:80->10.10.0.1:12345
+					* CT zone SNAT 65521          10.10.0.3:80->10.10.0.1:12345            =>  10.10.0.3:80->192.168.77.1:12345
+					* CT zone DNAT 65520          10.10.0.3:80->192.168.77.1:12345         =>  192.168.77.1:30001->192.168.77.1:12345
+					* output
+				  * packet 3 (request)
+					* client                     192.168.77.1:12345->192.168.77.100:30001
+					* CT zone SNAT 65521         192.168.77.1:12345->192.168.77.100:30001
+					* CT zone DNAT 65520         192.168.77.1:12345->10.10.0.3:80
+					* CT zone SNAT 65521         192.168.77.1:12345->10.10.0.3:80          =>  10.10.0.1:12345->10.10.0.3:80
+					* output
+				  * packet ...
+			*/
+			// The source IP address of packet 3 cannot be transformed after passing through zone 65521 (SNAT CT zone)
+			// since there is no conntrack record about 192.168.77.1:12345<->192.168.77.100:30001, and the source IP is
+			// still 192.168.77.100. As a result, subsequent request packets like packet 3 need to pass through SNAT CT
+			// zone to perform SNAT.
```

Comment:
This may be more clear: "As a result, subsequent request packets like packet 3 will only perform SNAT when they pass through SNAT CT zone the second time, after they are DNATed in DNAT CT zone."

Reply:
Done.